src/main/ngapp/accio/src/app/editor/magic-wand.service.ts

```diff
@@ -175,4 +201,130 @@ export class MagicWandService {
   }
+
+
+  /* Smarter flood fill tool */
+
+  scribbleFloodfill(imgData: ImageData, xCoord: number, yCoord: number,
+      tolerance: number, scribbles: Set<number>): Set<number> {
+    // Stores a queue of coords for pixels that we need to visit in "visit".
+    const visit: Array<Array<number>> = new Array();
+    // Stores already-visited pixels in "visited" as index formatted numbers
+    // (as opposed to coord format; for Set funcs).
+    const visited: Set<number> = new Set();
+    // Uses a set for mask; mainly do iter and set operations on masks.
+    const mask: Set<number> = new Set();
+    // Represents [R,G,B] attributes of initial pixel.
+    const originalPixel: Array<number> =
+        this.dataArrayToRgba(imgData, xCoord, yCoord);
+
+    visit.push([xCoord, yCoord]);
+    // Converts [x,y] format to 1-D equivalent of imgData.data (DataArray).
+    const indexAsDataArray: number =
+        this.coordToDataArrayIndex(xCoord, yCoord, imgData.width);
+    visited.add(indexAsDataArray);
+
+    // Filters scribbles by removing unnecessary pixels
+    scribbles = this.filterScribbles(scribbles, originalPixel, imgData, tolerance);
+
+    // Loops until no more adjacent pixels within tolerance level.
+    while (visit.length !== 0) {
+      const coord: Array<number> = visit.pop();
+      // Unpacks coord.
+      const x: number = coord[0];
+      const y: number = coord[1];
+
+      // Operational part of while-loop...
+      mask.add(this.coordToDataArrayIndex(x, y, imgData.width));
+
+      // Loop part of while-loop...
+
+      // Gets coords of adjacent pixels.
+      const neighbors: Array<Array<number>> =
+          [[x - 1, y], [x + 1, y], [x, y - 1], [x, y + 1]];
+      // Adds coords of adjacent pixels to the heap.
+      for (const neighborPixel of neighbors) {
+        const x: number = neighborPixel[0];
+        const y: number = neighborPixel[1];
+        // Checks if coord is in bounds and has not been visited first.
+        if (!this.getIsValid(imgData.width, imgData.height, x, y, visited)) {
+          continue;
+        }
+        // Visits the pixel and check if it should be part of the mask.
+        visited.add(this.coordToDataArrayIndex(x, y, imgData.width));
+        if (this.getIsScribbleMask(scribbles, imgData, neighborPixel, tolerance)) {
+          visit.push(neighborPixel);
+        }
+      }
+    }  // End of while loop...
+
+    return mask;
+  }
+
+  /**Judges current pixel's RGB against a set of reference pixels' RGBs to
+   * see if it can still be part of the mask (using tolerance criteria).
+   */
+  getIsScribbleMask(scribbles: Set<number>, imgData: ImageData,
+      pixelCoord: Array<number>, tolerance: number): boolean {
+    const curX: number = pixelCoord[0];
+    const curY: number = pixelCoord[1];
+
+    // Gets array of color attributes of current pixel.
+    const curPixel: Array<number> = this.dataArrayToRgba(imgData, curX, curY);
+
+    for (let pixelIndex of scribbles) {
```

Comment:
Would scribbles be better passed in as a set of (x, y) pairs? That way you don't have to translate them back.

Reply:
Nope, I don't mind; it's all good since you already have them in Set<number> format on your side, I was able to simply handle it with a helper function on my end that converts DataArray indexing into [x, y] format.